vendor/github.com/xtaci/smux/session.go

```diff
@@ -281,73 +281,91 @@ func (s *Session) recvLoop() {
 	}
 }
 
-func (s *Session) keepalive() {
-	tickerPing := time.NewTicker(s.config.KeepAliveInterval)
-	tickerTimeout := time.NewTicker(s.config.KeepAliveTimeout)
-	defer tickerPing.Stop()
-	defer tickerTimeout.Stop()
+// keepAliveSend will periodically send a keepalive meesage to the remote peer.
+func (s *Session) keepAliveSend() {
+	// Use 'time.Ticker' to ensure that if the keepalive sending falls behind
+	// schedule, unnecessary sleeping does not occur.
+	ticker := time.NewTicker(s.config.KeepAliveTimeout)
 	for {
 		select {
-		case <-tickerPing.C:
-			s.writeFrame(newFrame(cmdNOP, 0))
-			s.notifyBucket() // force a signal to the recvLoop
-		case <-tickerTimeout.C:
-			if !atomic.CompareAndSwapInt32(&s.dataReady, 1, 0) {
-				s.Close()
-				return
-			}
 		case <-s.die:
 			return
+		case <-ticker.C:
+			s.writeFrame(newFrame(cmdNOP, 0), time.Now().Add(s.config.WriteTimeout))
+			s.notifyBucket() // force a signal to the recvLoop
 		}
 	}
 }
 
-func (s *Session) sendLoop() {
-	buf := make([]byte, (1<<16)+headerSize)
+// keepAliveTimeout will periodically check that some sort of message has been
+// sent by the remote peer, closing the session if not.
+func (s *Session) keepAliveTimeout() {
+	// Use 'time.After' to ensure that at least 'n' amount of time passes
+	// between each check that a keepalive has been sent.
+	timeoutChan := time.After(s.config.KeepAliveTimeout)
```

Comment:
the loop in this function isn't doing any blocking work (just a single atomic op) so a ticker should function identically here